### PR TITLE
ADR-07: Full ABP-style DNSBL list support (DNS-only, in the Python build)

### PR DESCRIPTION
**Pull request by andrebrait**
Implements **ADR-07 — Full ABP-style DNSBL list support (DNS-only, in the Python build)**, all 9 phases, on `adr/07` (cut from `next`; depends on ADR-06 having landed).

## What this adds

DNS-only AdBlock-Plus / AdGuard filter-syntax support in the Unbound Python plugin: block `||domain^` / hosts / plain, `@@` allow exceptions, `/regex/` (reducible folds to dict form; irreducible compiled), `$options` (keep DNS-relevant, skip element/path/page-context), `$important` precedence inversion, feed-only `$badfilter` prune, and a 6-band provenance/sovereignty precedence model (user always wins over feeds). Stdlib-only, no subprocess, no query-time timeout.

## Phases (one commit each, gated green at every step)

1. `780d46a` — inventory + ABP corpus + regex/ReDoS measurement spike (de-risk, GO)
2. `a20cd7d` — ABP DNS decision spec + golden oracle (pure reference semantics)
3. `1097fe7` — matcher strata refactor + payload widening (behaviour-preserving)
4. `d384252` — Rule model + DNS-only ABP Stage-A parser
5. `2267c92` — Stage-B reconcile (`$badfilter` prune + regex reduction + bands)
6. `d0decde` — emit + wire ABP build + **live** 6-band precedence matcher
7. `914459c` — regex safety (opt-in static cap + runtime warn/evict, feed + user)
8. `8fe250f` — slim PHP to abp-tag + raw passthrough; delete lite parser; keep DNSBL-IP
9. `a7c6f5b` — validation, perf/ReDoS evidence, docs, smoke checklist + DoD

## Verification

- `python -m pytest` → **917 passed** (ADR-06 no-regression oracle + the full ADR-07 ABP suites all green).
- `ruff check`/`ruff format`, `php -l`, markdownlint all clean.
- Benchmark (reproduces Phase-1, GATE = GO): reduction 46% (≥30%), 0.165 µs/irreducible-pattern (~303-pattern budget at 50 µs), ReDoS static cap 1 caught / 0 missed, worst cap-passing first-hit 0.109 ms ≤ 500 ms.

## Status

`Implemented — pending live smoke` — **not** Accepted. No live Unbound in CI; the maintainer must run the §7 manual smoke on a live pfSense box (after a full DNSBL update + reload) before flipping to Accepted. Checklist is in `.ADRs/ADR_07_ABP_DNSBL_Support/RESULTS/09_Results.txt`:

1. `@@` exception un-blocks a carve-out subdomain; rest of zone stays blocked.
2. Regex: reducible == dict form; irreducible blocks; `@@/re/` un-blocks; `DNSBL_Regex` alias reflects admitted count post-reload.
3. `$important` beats feed `@@`; `$badfilter` removes a feed block; neither touches a user rule.
4. User sovereignty: whitelist (textarea + alerts button), user block beats feed `@@$important`, TOP1M as user allow.
5. Regex safety: cap ON drops over-long feed/user regex at load; a slow regex warns then evicts and the resolver recovers; confirm `time.thread_time` resolves on FreeBSD.
6. DNSBL-IP: `||1.2.3.4^`/hosts IPs populate `DNSBLIP_v4/_v6`, no IP leak; `.abp` marker round-trips a reuse build.
7. No regression on plain/hosts feeds; `pfb_py_count` renders.
8. Reload picks up feed/whitelist/regex-list/setting changes.

Accepting ADR-07 unblocks the parked ADR-08 (IDN homoglyph protection).

---
_Generated by [Claude Code](https://claude.ai/code/session_01X7BdgA5T3hRnxsJ2rH1Z4N)_

## Summary by CodeRabbit

* **New Features**
  * Added full Adblock Plus (ABP) / EasyList DNSBL feed support with exception handling (`@@` rules) and regex-based blocking.
  * Added regex safety protections: optional static complexity cap at load time and always-on runtime monitoring with warnings and automatic eviction of pathological patterns.
  * Added "Limit long/complex regex" configuration option in the DNSBL settings.
  * Enhanced user domain sovereignty: custom domain lists and user-defined regex rules now protected from feed precedence rules.

* **Documentation**
  * Expanded documentation for ABP/EasyList feed processing and regex safety behavior.

### CODE REVIEW

**Comment by andrebrait** [comment]
@coderabbit trigger full review

---
_Generated by [Claude Code](https://claude.ai/code)_

File: benchmarks/spike_adr07_regex.py [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+
+from __future__ import annotations
+
+import os
+import re
+import statistics
+import sys
+import threading
+import time
+from typing import Callable
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply benchmarks/spike_adr07_regex.py]: **Applied** in b3b2d45 — removed the unused `Callable` import.

(For the record: `F401` is in this repo's ruff `ignore` list — "some imports are conditional/side-effect" — so CI was green regardless of Flake8 flagging it. But the import is genuinely dead, so it's gone.)

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: benchmarks/spike_adr07_regex.py [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+# the resolver, fact 2 means this exact match CANNOT be killed, which is the point).
+_CHILD = (
+    "import re,sys,time;"
+    "p=sys.stdin.read();"
+    "rx=re.compile(p);"
+    "s='a'*%d;"
+    "t=time.perf_counter();"
+    "rx.search(s);"
+    "sys.stdout.write(repr(time.perf_counter()-t))" % DNS_MAX_LEN
+)
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply benchmarks/spike_adr07_regex.py:249]: **Applied** in b3b2d45 — and this turned out to be more than a benchmark nit. `_adversarial_input` was dead code and `_CHILD` hard-coded `'a'*253`, which *matches* `^(a+)+$` / `^(a|a)+$` instead of failing the final anchor. With the corrected failing probe (`'a'*(N-1)+'!'`, now passed to the child via JSON), the gate flipped to **NO-GO**: `^(a|a)+$` **passes** the static cap (the nested-quantifier heuristic only caught a quantifier *inside* a group, not a quantified alternation) yet backtracks catastrophically — and since `re.search` is uninterruptible (fact 2), the runtime evict can't kill it in-flight.

Fixed by strengthening the cap heuristic to also flag alternation-overlap (`(...|...)[+*{]`) in **both** the spike and production `_regex_exceeds_static_cap`. Benchmark is back to **GO** (pathological caught by cap: 4 / missed: 0). ADR §7 evidence updated, and a non-blocking follow-up was logged for the cap-default + unkillable-in-flight-ReDoS question.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/usr/local/pkg/pfblockerng/pfb_unbound.py [hunk 3]
```diff
@@ -87,6 +87,7 @@
 dataDB: defaultdict[str, Any]
 zoneDB: defaultdict[str, Any]
 regexDB: defaultdict[str, Any]
+allowRegexDB: defaultdict[str, Any]
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply src/usr/local/pkg/pfblockerng/pfb_unbound.py:90]: **Skipping** — `allowRegexDB` is annotation-only at module scope *by design*, exactly like every other DB global (`regexDB`, `dataDB`, `zoneDB`, …). The block comment there states they're "Declared here without assignment (PEP 526) … no runtime object is created" and are populated by `init_standard()` at runtime (`allowRegexDB` is assigned under the `global` block, ~line 408) before `operate()` ever reads it.

Initialising only `allowRegexDB` would break consistency with the established pattern without fixing a reachable `NameError` (`operate()` never runs pre-init). ruff doesn't flag annotation-only globals, and CI is green.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/usr/local/pkg/pfblockerng/pfb_unbound.py [hunk 4]
```diff
@@ -1780,6 +1946,241 @@ def _dnsbl_parse_abp_line(line: str) -> str | None:
...
+            key_or_pattern=inner,
+            important=False,
+            badfilter=False,
+            provenance=provenance,
+            feed=feed,
+            group=group,
+            log=log,
+            signature=(inner, ()),
+            wildcard=False,
+        )
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply src/usr/local/pkg/pfblockerng/pfb_unbound.py]: **Applied** in 03578b8 — with the spec kept in lockstep. Worth noting: production `parse_abp` matched the *ratified* Phase-2 oracle (`parse_abp_line`), which also only accepted bare `/re/` and `@@/re/` — so this was a deliberate scope boundary, not a prod-vs-spec drift. Per maintainer decision we've now extended **both**: regex rules accept an optional `$options` suffix, parsed via the existing `_dnsbl_classify_options` (page-context / non-DNS / unrecognised → skip the whole rule), setting `important` / `badfilter` and `signature=(inner, dns_opts)`. reconcile's `$badfilter`-prune-by-signature and `$important` band-raise now apply to regex rules too. `prod == oracle` verified across the corpus; parser pins + live emit tests added.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/usr/local/pkg/pfblockerng/pfb_unbound.py [hunk 5]
```diff
@@ -2018,43 +2858,98 @@ def index_for(feed: str, group: str) -> int:
...
+            else:
+                data_db[b.key] = payload
+
+        # Domain allows (@@||domain^ and reduced @@/re/) -> whiteDB. A user whitelist
+        # entry (band 6) must never be downgraded by a feed allow on the SAME key, so
+        # only write when the key is absent or the existing band is lower.
+        for a in result.allow_domains:
+            existing = white_db.get(a.domain)
+            if existing is None or _white_entry_band(existing) < a.band:
+                white_db[a.domain] = {"wildcard": a.wildcard, "important": a.important, "band": a.band}
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply src/usr/local/pkg/pfblockerng/pfb_unbound.py:2995]: **Applied** in 1f071e5 — confirmed reachable. reconcile appends one `AllowDomainRule` per surviving allow (no dedup), so a single key can carry both an exact (reduced `@@/^d$/`) and a wildcard (`@@||d^`) allow at the **same band**, and the whitelist suffix-walk only honours *wildcard* entries. First-writer-wins would drop subdomain coverage order-dependently. It now merges monotonically on a same-band collision (`wildcard`/`important` = OR, `band` = max). Regression test added: exact-allow-emitted-first still keeps `sub.example.com` covered.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/usr/local/pkg/pfblockerng/pfblockerng.inc [hunk 6]
```diff
@@ -2315,22 +2317,78 @@ function pfb_dnsbl_whitelist_lines() {
...
+			return $host;
+		}
+		return '';
+	}
+
+	// hosts line: "<sink-ip> <target>" where the target is itself an IP.
+	if (strpos($s, ' ') !== FALSE) {
+		$parts = preg_split('/\s+/', $s);
+		if (is_array($parts) && count($parts) >= 2 &&
+		    (is_ipaddrv4($parts[0]) || is_ipaddrv6($parts[0]))) {
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply src/usr/local/pkg/pfblockerng/pfblockerng.inc]: **Applied** in 1f071e5 — the gate is now `preg_match('/\s/', $s)`, so tab-delimited hosts lines reach the IP path (the following `preg_split('/\s+/')` already split on any whitespace). The Python boundary oracle mirrors this, and a tab-delimited hosts case was added to the coexistence test.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/usr/local/pkg/pfblockerng/pfblockerng.inc [hunk 7]
```diff
@@ -7820,16 +7907,53 @@ function sync_package_pfblockerng($cron='') {
...
 												}
 
-												$lite = TRUE;
-												$line = str_replace($e_replace, '', $line);
+												// Pass the RAW ABP line through to Python verbatim
+												// (NO domain extraction). The manifest writer copies
+												// 'abp' feed .txt lines into the per-feed raw as-is.
+												@fwrite($dhandle, "{$line}\n");
+												$domain_data = "{$line}\n";
+												continue;
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply src/usr/local/pkg/pfblockerng/pfblockerng.inc:8008]: **Applied** in af247d0 — on the consuming side. The Phase-8 manifest writer (`pfb_unbound_python_sources`) runs unconditionally, and `dnsbl_build_from_manifest()` / `build()` is the primary path that parses ABP feeds via `parse_abp` *and* does its own TLD classification — so the legacy data/zone files `tld_analysis()` writes aren't read once the manifest build succeeds. The genuine risk was `tld_analysis()` CSV-mangling ABP raw lines under TLD mode, so it now **skips any feed carrying the `.abp` marker** (and empty-feed-column raw lines), gated on `!empty($abp_feeds)` so plain-only configs are byte-identical. ABP feeds still build via Python; the initial ABP × TLD-mode behaviour is documented in README + ADR with a **follow-up** to review full integration, plus a smoke-checklist item.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/usr/local/pkg/pfblockerng/pfblockerng.inc [hunk 8]
```diff
@@ -8307,10 +8443,29 @@ function sync_package_pfblockerng($cron='') {
...
+		if (file_exists($pfb['unbound_py_regex_count'])) {
+			$raw_rc = trim((string) @file_get_contents($pfb['unbound_py_regex_count']));
+			if ($raw_rc !== '' && ctype_digit($raw_rc)) {
+				$py_regex_cnt = (int) $raw_rc;
+			}
+		}
+		if ($pfb['dnsbl_regex'] == 'on' || ($py_regex_cnt !== NULL && $py_regex_cnt > 0)) {
+			if ($py_regex_cnt !== NULL) {
+				$regex_cnt = $py_regex_cnt;
+			} else {
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply src/usr/local/pkg/pfblockerng/pfblockerng.inc:8517]: **Skipping** — this is documented, intentional best-effort (see the block comment directly above the code): the count file is written at resolver (re)load, so the pass falls back to the configured user-regex line count when it's absent/stale — same nature as `pfb_py_count`. During a pending reload the file still holds the *currently-active* admitted count, which is arguably a better value for the alias than the user-line-count fallback; clearing it pre-reload isn't clearly more correct. It's already called out for the maintainer's live smoke (ADR §7), so I'd rather not pre-empt that validation with a behavioural change here.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_adr07_decision_spec.py [hunk 9]
```diff
@@ -0,0 +1,862 @@
...
+                target=Target.DOMAIN,
+                key=dom,
+                provenance=provenance,
+                wildcard=wildcard,
+                signature=(dom, ()),
+            )
+        try:
+            re.compile(inner)
+        except re.error:
+            return None
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply tests/test_adr07_decision_spec.py:324]: **Skipping** — no net-decision divergence on any reachable input. The oracle is the all-in-one parse+decide spec, so it validates regex eagerly; production splits Stage-A parse (raw, deferred) from compile-at-load. A malformed/uncompilable `/re/` ends up **inactive either way**: the oracle drops it at parse, production drops it when `_dnsbl_compile_regex_rules` fails to compile at load — same decision. The production parser's raw-defer contract is already pinned by `tests/test_adr07_parser.py`. Removing the oracle's eager compile would perturb the ratified spec for no decision change. (The regex+`$options` work in 03578b8 kept `prod == oracle`.)

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_adr07_php_boundary.py [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+  2. no firewall IP leaks into the domain/regex build, and the IP-anchored lines
+     land in the firewall set instead, and
+  3. the boundary is RAW: a ``$``/``@@``/``/re/`` line reaches Python intact (it
+     would have been dropped by the old PHP lite pass).
+"""
+
+from __future__ import annotations
+
+import re
+from typing import Any, Iterable
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply tests/test_adr07_php_boundary.py]: **Applied** in 1f071e5 — `_php_abp_extract_ip` now uses `ipaddress.ip_address` (mirroring PHP `is_ipaddrv4`/`is_ipaddrv6`) instead of the IPv4-only regex, and the coexistence FEED covers ABP-anchored + hosts IPv6 (`||2001:db8::7^`, `:: 2001:db8::99`) plus a tab-delimited hosts line — all asserted into the firewall set.

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by andrebrait** [comment]
## CodeRabbit review — resolved

Validated all 11 findings against the current code. Summary:

**Applied (6):**
- **Unused `Callable` import** — removed (`b3b2d45`).
- **ReDoS benchmark probe** — corrected (`b3b2d45`); this exposed a real gap (`^(a|a)+$` alternation-overlap passed the static cap), so the cap heuristic was strengthened in both the spike and production; benchmark back to **GO** (4 pathological caught / 0 missed).
- **`$important`/`$badfilter` on regex rules** — parser + ratified Phase-2 oracle extended in lockstep (`03578b8`).
- **whiteDB same-band exact+wildcard merge** — now monotonic-widen, regression test added (`1f071e5`).
- **Tab-delimited hosts lines** — whitespace gate (`1f071e5`).
- **IPv6 DNSBL-IP boundary coverage** — `ipaddress`-based + IPv6/tab cases (`1f071e5`).

**Applied — coexistence (1):**
- **ABP feeds × DNSBL TLD mode** — `tld_analysis()` now skips `.abp`-marked feeds so their raw lines aren't CSV-mangled; they build via the Python manifest path. Documented + follow-up logged (`af247d0`).

**Skipped (3, with reasons on the threads):**
- **`allowRegexDB` module init** — annotation-only is the deliberate, documented pattern for *all* DB globals; populated in `init` before `operate()` runs.
- **Stale `unbound_py_regex_count`** — documented best-effort fallback (same as `pfb_py_count`); flagged for the maintainer's live smoke.
- **Oracle eager-compile vs Stage-A** — no net-decision divergence (uncompilable regex is inactive either way); the raw-defer contract is already pinned by `test_adr07_parser.py`.

### Nitpick — `RESULTS/07_Results.txt:180` ("user admitted count" → "user-admitted count")

**Skipping.** The `.ADRs/**/RESULTS/*.txt` files are verbatim phase-handoff records (the interface between ADR phases), not living documentation — they aren't retro-edited after a phase lands. No functional impact.

Suite is green (934 passed), ruff/php-l/markdownlint clean. Two non-blocking follow-ups were recorded in `ADR.md` (regex-cap default + unkillable in-flight ReDoS; full ABP×TLD integration).

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by andrebrait** [comment]
## Status update — rebased on `next`, smoke testing added, sovereignty fixes, pending live smoke

### Rebased onto current `next` (force-pushed)
`adr/07` was rebased onto the current `next` tip (the 3 already-upstream plan-doc commits
dropped; the 24 ADR-07 commits replayed clean). History was rewritten — **compare by
content, not hash.** Verified sound: full suite green together with next's ADR-06 +
build-pkg tests; next's python-mode smoke harness, chroot feed-load fix, and
build-pkg-portable chain are all inherited. One post-rebase consistency fix: dropped a
stray `(py)` marker on the new regex-cap field (next made the DNSBL UI python-only).

### Smoke testing (the purpose of this round)
- **`tests/smoke/test_smoke_abp.py`** — ABP decision matrix on the live-VM harness:
  `@@` un-block (same feed + cross-feed), `$important`/`$badfilter` precedence, regex
  block + `@@` allow + admitted-count-under-cap, whitelist sovereignty, user-regex
  sovereignty, Custom_List sovereignty, plain-feed no-regression.
- **Harness extensions** (`helpers.py`): `abp_feed()`, `DnsblCase` ABP fields
  (`extra_rows`, `user_regex`, `regex_cap`, `custom_domains`, `cname_validation`),
  Python-emitted count readers — plus pure (no-VM) self-tests pinning the inject snippets.
- **`ADR.md`** "Manual smoke" → "Live smoke" section mapping each checklist item to its
  automated test or a documented hand-run reason; `RESULTS/` scaffold added.

### Sovereignty fixes (surfaced while writing the smoke)
Fact 7 ("user intent is sovereign") was under-implemented:
- **User regex** loaded as a bare pattern → feed band 1, so a feed `@@…$important`
  (band 4) could override it. Now loaded as the band-5 `PRIO_USER_BLOCK` payload.
- **DNSBL Group Custom_List** domains entered the build as a feed row (band 1). Now
  tagged `provenance='user'` through the manifest → band 5. Both beat any feed allow;
  only the band-6 whitelist overrides. Unit-pinned (`TestUserRegexSovereign`,
  `TestUserCustomListSovereign`, incl. the no-user-tag contrast).

### CNAME validation
Unit coverage added for the with/without/direct matrix
(`TestOperateDnsbl::test_cname_*`); the live A→CNAME→B smoke is a documented skip
pending controlled-upstream-CNAME harness support (tracked: #41).

### Status / gate
**`Implemented — pending live smoke`.** The decision logic is fully unit-covered; the
end-to-end gate is a live VM dispatch of `smoke.yml` (I can't run it from CI here). The
ADR's "Live smoke" matrix lists which items are automated vs hand-run (runtime regex
evict, ABP×TLD live, async DNSBL-IP populate — the last shares #35's deferral). Flip to
**Accepted** only after that passes + `RESULTS/` recorded.

_Out-of-scope discoveries from this round were routed elsewhere, not bundled here:_
cache/reporting findings (#26, #43, #44), kill-gate enforcement (#42), coverage tooling
(#38–#40), and ADR-10 doc corrections (on `next`).

**Comment by andrebrait** [comment]
@coderabbitai review

Please run a full review of this PR and tell me when you're finished. (Auto-review is skipped because the base branch is `next`, our development branch, not the default `devel`.)

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

**Comment by andrebrait** [comment]
@coderabbitai review

Rebased onto the latest `next` and force-pushed (now at `55e5e1b`). Please run a full review of the current HEAD and tell me when you're finished. (Auto-review is skipped because the base is `next`, our development branch.)

---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

File: .ADRs/ADR_07_ABP_DNSBL_Support/RESULTS/01_Results.txt [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+negative query, so the ~50 us added-latency budget bounds to ~300 irreducible
+patterns — comfortably above any realistic per-deployment count once anchored
+patterns fold to dicts (measured 46% reduction on a metachar-heavy corpus; real
+DNS feeds fold far higher because most /re/ are simple anchored-domain forms).
+The accepted ReDoS residual is real but bounded: a static nested-quantifier+length
+cap catches the genuinely catastrophic shapes (the only pattern that ran >2 s in
+the corpus was nested-quantifier-flagged), and every cap-passing pattern's worst
+first-hit on a <=253-char DNS name was <0.15 ms. The GIL demo confirms a
+thread/asyncio timeout CANNOT cancel a runaway match — which is exactly why the
+design accepts one slow first-hit then EVICTS, rather than vetting.
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply .ADRs/ADR_07_ABP_DNSBL_Support/RESULTS/01_Results.txt:18]: **Skipping** — this is a historical Phase-1 *handoff* snapshot (`RESULTS/01`). The alternation-overlap arm was added later (b3b2d45) and is recorded with that change; the production static cap + its unit tests (`test_adr07_regex_safety`) are correct. Backfilling a completed-phase handoff to describe a later phase's heuristic would misrepresent the record, and the "later phases re-implement the weaker heuristic" risk is moot now that all phases shipped. Happy to reverse this if you'd rather keep the RESULTS docs as living records.


---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

File: .ADRs/ADR_07_ABP_DNSBL_Support/RESULTS/04_Results.txt [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+    domain-vs-regex, NOT kind).
+  RULE_TARGET_DOMAIN="domain"  RULE_TARGET_REGEX="regex"
+  RULE_PROV_USER="user"  RULE_PROV_FEED="feed"
+
+SIGNATURE DEFINITION (the $badfilter match key, Phase 5 uses it):
+  signature = (key_or_pattern, tuple(sorted DNS-options))  MINUS $badfilter.
+  * DOMAIN rule: (domain, dns_opts) where dns_opts is the SORTED tuple of
+    DNS-relevant option NAMES excluding badfilter -> ("important",) or ().
+  * REGEX rule: (inner_pattern, ()).  ($options on a /re/ are not parsed -- regex
+    rules carry no options in this grammar.)
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply .ADRs/ADR_07_ABP_DNSBL_Support/RESULTS/04_Results.txt:60]: **Skipping** — historical Phase-4 handoff snapshot (`RESULTS/04`). Regex `$options` support landed in a later commit; production `parse_abp` and the oracle handle it and are tested. Skipping the backfill of a completed-phase record (same offer: I'll update the RESULTS docs if you prefer them living).


---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

File: .ADRs/ADR_07_ABP_DNSBL_Support/RESULTS/05_Results.txt [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+    * any rule with $important (feed block/allow + $important), OR
+    * any FEED ALLOW (@@) -- a feed exception the fast path's "allow beats block"
+      seam handles only via whiteDB; with bands it needs numeric resolution, OR
+    * any FEED irreducible regex (block or allow).
+It is FALSE for a pure plain/hosts/||domain^ FEED-block set (today's corpus) AND
+for a USER-only rule set (the fast path already handles user whitelist + user
+blocks; P3 SS3) -- pinned by TestImportantRulesFlag (5 cases). So:
+  PHASE 6: emit pfb["important_rules"] = ReconcileResult.important_rules, and
+  pfb["allowRegexDB"] = bool(allow_regex_irreducible OR any allow_domains). When
+  important_rules is False the matcher stays byte-for-byte today (ADR-06 oracle).
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply .ADRs/ADR_07_ABP_DNSBL_Support/RESULTS/05_Results.txt:172]: **Skipping** — historical Phase-5 handoff snapshot (`RESULTS/05`). The runtime contract (allow domains → `whiteDB`; `allowRegexDB` reflects admitted allow-regex only) is correct in production and tested; the stale prose is a point-in-time snapshot. Skipping the backfill (offer stands to update if you want living docs).


---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

File: benchmarks/spike_adr07_regex.py [hunk 14]
```diff
@@ -0,0 +1,507 @@
...
+    if not s or s.startswith("!") or s.startswith("["):
+        return "comment"
+    # element-hiding / cosmetic
+    if "##" in s or "#@#" in s or "#?#" in s:
+        return "element_hiding"
+    # regex rules
+    if s.startswith("@@/") and s.endswith("/"):
+        return "regex_allow"
+    if s.startswith("/") and s.endswith("/") and len(s) > 2:
+        return "regex_block"
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply benchmarks/spike_adr07_regex.py:88]: **Valid latent gap — tracked in #47.** `categorise()`/`_strip_slashes()` predate the regex-`$options` support, so `/foo/$important` / `@@/foo/$badfilter` would be misclassified/mis-stripped. The spike is the informational Phase-1 reference (always exits 0; enforcement is #42), and the current corpus doesn't route `/re/$options` through it, so there's no live or test impact today. Noted in #47 to fix if such fixtures are added, so the reference can't drift from the oracle/production.


---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

File: src/usr/local/pkg/pfblockerng/pfb_unbound.py [hunk 15]
```diff
@@ -1780,6 +1960,267 @@ def _dnsbl_parse_abp_line(line: str) -> str | None:
...
+    # ---- hosts: "<ip> <domain>" ----------------------------------------- #
+    if " " in s:
+        first, _, target = s.partition(" ")
+        target = target.strip()
+        if not _dnsbl_is_ipv4(first):
+            return None  # not a hosts line (a real ABP line never has a bare space)
+        if _dnsbl_is_ipv4(target):
+            return None  # "<ip> <ip>" -> firewall path
+        dom = normalise(target)
+        if dom is None:
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply src/usr/local/pkg/pfblockerng/pfb_unbound.py:2190]: **Applied** in 55bb589. `parse_abp` now splits on any whitespace (`s.split(None, 1)`), so tab-delimited hosts lines (`0.0.0.0	domain`) parse exactly like space-delimited ones instead of falling through to the bare-domain path. Added `test_adr07_parser.py::test_hosts_tab_delimited` (incl. `<ip>	<ip>` → firewall path). Thanks — good catch.


---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

File: src/usr/local/pkg/pfblockerng/pfb_unbound.py [hunk 16]
```diff
@@ -2248,6 +3278,194 @@ class DnsblDecision:
...
+            pattern = r.get("re") if isinstance(r, dict) else r
+            match, elapsed_ms = _regex_timed_search(pattern, q_name)
+            if _regex_should_evict(_k, elapsed_ms, warn_ms, evict_ms, "DNSBL_Regex", _k):
+                to_evict.append(_k)
+                continue
+            if match:
+                best = max(best, _block_entry_band(r))
+        if to_evict:
+            _regex_evict_names(regex_db, to_evict)
+    return best
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply src/usr/local/pkg/pfblockerng/pfb_unbound.py:3371]: **Valid — deferred to #47.** Confirmed: `_scan_block_band` returns only the band, so `feed`/`group`/`log_type`/`b_eval` stay bound to the first-discovered block; a higher-band winner (e.g. a user regex beating a feed block on the same name) gets the right priority but the wrong response shape + attribution. It's an edge case (same name matched at multiple bands) and a heavier refactor (change the helper's return + both call sites + tests), so it's tracked in #47 rather than expanded into this PR. Side note: I also fixed the stale "UNREACHABLE in production" comment near here — the numeric branch IS wired now (ADR-07 P6).


---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

File: src/usr/local/pkg/pfblockerng/pfblockerng.inc [hunk 17]
```diff
@@ -2949,6 +3052,17 @@ function tld_analysis() {
...
+			// ADR-07 P8: skip ABP-feed lines (see note above). A plain CSV row is
+			// ',domain,,log,feed,group'; the feed name is column index 4. ABP raw
+			// lines lack this shape, so an unset/empty feed column also means skip.
+			if (!empty($abp_feeds)) {
+				$fparts	= explode(',', $line, 6);
+				$lfeed	= isset($fparts[4]) ? $fparts[4] : '';
+				if ($lfeed === '' || isset($abp_feeds[$lfeed])) {
+					continue;
+				}
+			}
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply src/usr/local/pkg/pfblockerng/pfblockerng.inc:3064]: **Real but narrow — deferred.** The `$lfeed === ''` clause already skips the common ABP shapes (`||x^`/`@@`/hosts/plain have <5 commas → empty feed column → skipped); only a raw ABP **regex** line with ≥5 commas slips through, and only under DNSBL-TLD mode. That's the ABP×DNSBL-TLD coexistence path already carrying the `FOLLOW-UP` marker just above this code (the proper fix is provenance-preserving merge / folding the TLD pass into the Python build). It's also a hand-run item in the ADR-07 live-smoke matrix. Left for that focused pass rather than a fragile comma-count tweak.


---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

File: tests/test_adr07_decision_spec.py [hunk 18]
```diff
@@ -0,0 +1,915 @@
...
+        sig_opts = tuple(o for o in dns_opts) + (("important",) if important else ())
+        return Rule(
+            kind=kind,
+            target=Target.REGEX,
+            key=inner,
+            provenance=provenance,
+            important=important,
+            badfilter=badfilter,
+            options=dns_opts,
+            signature=(inner, tuple(sorted(sig_opts))),
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply tests/test_adr07_decision_spec.py:334]: **Applied** in 55bb589. The reduced-DOMAIN branch now keeps `signature=(inner, …)` instead of the reduced domain, so a literal `||domain^$badfilter` can't prune a reducible regex (and vice-versa). This aligns the oracle with production `parse_abp` (`signature=(inner, dns_opts)`, :2137); equivalence tests stay green (the divergence was latent — no fixture exercised it).


---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.

File: tests/test_adr07_parser.py [hunk 19]
```diff
@@ -0,0 +1,442 @@
...
+# --------------------------------------------------------------------------- #
+
+
+def _spike():
+    spike_path = os.path.join(os.path.dirname(__file__), "..", "benchmarks", "spike_adr07_regex.py")
+    s = importlib.util.spec_from_file_location("spike_adr07_regex", spike_path)
+    assert s and s.loader
+    mod = importlib.util.module_from_spec(s)
+    s.loader.exec_module(mod)
+    return mod
```
[the remark this answers is not in the document]
Comment (andrebrait) [reply tests/test_adr07_parser.py]: **Applied** in 55bb589 — `_spike() -> ModuleType` (+ `from types import ModuleType`).


---
🤖 Generated by [Claude Code](https://claude.com/claude-code), posted via @andrebrait's account on their behalf.
